src/eduid_userdb/idp.py

```diff
@@ -0,0 +1,196 @@
+                    'eduPersonScopedAffiliation',
+                    'givenName',
+                    'mail',
+                    'norEduPersonNIN',
+                    'preferredLanguage',
+                    'sn',
+                    ]
+
+
+class IdPUser(User):
+    """
+    Wrapper class for eduid_userdb.User adding functions useful in the IdP.
+    """
+
+    def to_saml_attributes(self, config, logger, filter_attributes=_SAML_ATTRIBUTES):
+        """
+        Return a list of SAML attributes for a user.
+
+        Note that this is _all_ parts of the user that this IdP knows how to express as
+        SAML attributes. It is not necessarily the attributes that will actually be released.
+
+        :param config: IdP config
+        :param logger: logging logger
+        :param filter_attributes: Filter to apply
+
+        :type config: eduid_idp.config.IdPConfig
+        :type logger: logging.Logger
+        :type filter_attributes: [str | unicode]
+
+        :return: SAML attributes
+        :rtype: dict
+        """
+        attributes_in = self.to_dict(old_userdb_format = True)
+        attributes = {}
+        for approved in filter_attributes:
+            if approved in attributes_in:
+                attributes[approved] = attributes_in.pop(approved)
+        logger.debug('Discarded non-attributes:\n{!s}'.format(pprint.pformat(attributes_in)))
+        attributes1 = _make_scoped_eppn(attributes, config)
+        attributes2 = _add_scoped_affiliation(attributes1, config)
+        attributes = _add_eduperson_assurance(attributes2, self)
+        return attributes
+
+
+class IdPUserDb(object):
+    """
+
+    :param logger: logging logger
+    :param config: IdP config
+    :param userdb: User database
+
+    :type logger: logging.Logger
+    :type config: eduid_idp.config.IdPConfig
+    :type userdb: eduid_userdb.UserDB
+    """
+
+    def __init__(self, logger, config, userdb = None):
+        self.logger = logger
+        self.config = config
+        if userdb is None:
+            userdb = UserDB(config['MONGO_URI'], db_name=config['USERDB_MONGO_DATABASE'], user_class=IdPUser)
+        self.userdb = userdb
+
+    def lookup_user(self, username):
+        """
+        Load IdPUser from userdb.
+
+        :param username: string
+        :return: user found in database
+        :rtype: IdPUser | None
+        """
+        _user = None
+        if isinstance(username, string_types):
+            if '@' in username:
+                _user = self.userdb.get_user_by_mail(username.lower(), raise_on_missing=False)
+            if not _user:
+                _user = self.userdb.get_user_by_eppn(username.lower(), raise_on_missing=False)
+        if not _user:
+            # username will be ObjectId if this is a lookup using an existing SSO session
+            _user = self.userdb.get_user_by_id(username, raise_on_missing=False)
+        return _user
+
+
+def _make_scoped_eppn(attributes, config):
```

Comment:
I feel that this function and the ones below belongs in eduid_common but I wont push the issue if someone else feel otherwise :)

Reply:
I don't feel otherwise, I'll move them :)

Reply:
However, the leading underscore in their name suggests they should be private to the module they are defined in, I'll remove them